### PR TITLE
DDF-2195 Create Registry Rest Endpoint

### PR DESCRIPTION
**Pull request by gordocanchola**
#### What does this PR do?
This PR changes moves the registry-report-view module to a more generic registry-rest-endpoint to have the rest endpoint be more generic. The business logic of the report viewer was moved out of the endpoint and into a helper class. 
Publish and unpublish are added to the endpoint. 
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@lessarderic @clockard @ryeats 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@figliold 
@shaundmorris 
#### How should this be tested?
Verify the build
#### Any background context you want to provide?
The registry report viewer code was pretty much just moved into a helper method. That's the bulk of those changes. Most of the report viewer stuff is showing up because of the move
#### What are the relevant tickets?
DDF-2195
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

Changing the registry-report-view module to registry-rest-endpoint
Moving the buildRegistryMap logic to a RegistryReportHelper
Adding publish to the rest endpoint  POST /{registryId}/publication/{sourceId}
Adding unpublish to the rest endpoint DELETE /{registryId}/publication/{sourceId}
Updating tests

### CODE REVIEW

File: catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        assertThat(general.get("inputDataSources"), is("youtube, myCamera"));
+
+        // Collections
+        Map<String, Object> collections = getMapFromMap(registryMap, "Collections");
+        Map<String, Object> anotherName = getMapFromMap(collections, "Another Name");
+        assertThat(anotherName.get("parameters"), is("param1"));
+
+        // Services
+        Map<String, Object> services = getMapFromMap(registryMap, "Services");
+        //        Map<String, Object> serviceName = getMapFromMap(services, "Service Name");
```
Comment (vinamartin) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelperTest.java]: Woops, looks like somehow this got missed in my code review.

Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelperTest.java]: Woops on me. I commented that out and forgot to remove it. Thanks.

Comment (vinamartin) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelperTest.java]: Oh nice, so it wasn't my fault. :)

File: catalog/spatial/registry/registry-rest-endpoint/pom.xml [hunk 2]
```diff
@@ -50,11 +50,7 @@
         <dependency>
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-lang3</artifactId>
-            <version>3.4</version>
-        </dependency>
-        <dependency>
-            <groupId>org.codice.ddf.registry</groupId>
```
Comment (clockard) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: We still use RegistryPackageType in the endpoint so i think we still need the schema bindings

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        if (StringUtils.isBlank(registryId) || StringUtils.isBlank(destinationId)) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .entity("Registry and destination ID cannot be blank")
+                    .build();
+        }
+
+        try {
+            registryPublicationService.publish(registryId, destinationId);
+        } catch (FederationAdminException e) {
+            return Response.status(Response.Status.BAD_REQUEST)
```
Comment (clockard) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Should this be `Response.Status.INTERNAL_SERVER_ERROR`

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        if (StringUtils.isBlank(registryId) || StringUtils.isBlank(destinationId)) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .entity("Registry and destination ID cannot be blank")
+                    .build();
+        }
+
+        try {
+            registryPublicationService.unpublish(registryId, destinationId);
+        } catch (FederationAdminException e) {
+            return Response.status(Response.Status.BAD_REQUEST)
```
Comment (clockard) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Same comment about error type

File: catalog/spatial/registry/registry-rest-endpoint/pom.xml [hunk 5]
```diff
@@ -50,11 +50,7 @@
...
-            <artifactId>registry-schema-bindings</artifactId>
+            <version>3.0</version>
```
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: Seeing commons-lang3 used in a few places, we should consider doing dependency management on it. Why did you drop the version from 3.4 to 3.0?

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        templateLoader.setSuffix(".hbt");
+    }
+
+    @Path("/{registryId}/publication/{sourceId}")
+    @POST
+    public Response publish(@PathParam("registryId") String registryId,
+            @PathParam("sourceId") String destinationId) {
+        if (StringUtils.isBlank(registryId) || StringUtils.isBlank(destinationId)) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .entity("Registry and destination ID cannot be blank")
```
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: The message says registry **and** destination ID cannot be blank but the if statement checks **or**. So is the message correct or the if condition?

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                .build();
+    }
+
+    @Path("/{registryId}/publication/{sourceId}")
+    @DELETE
+    public Response unpublish(@PathParam("registryId") String registryId,
+            @PathParam("sourceId") String destinationId) {
+        if (StringUtils.isBlank(registryId) || StringUtils.isBlank(destinationId)) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .entity("Registry and destination ID cannot be blank")
```
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Since you are doing the same parameter validation of registry ID and destination ID in each of these methods you should extract that logic out to reduce the code duplication.

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                    .build();
+        }
+
+        RegistryPackageType registryPackage;
+        try {
+            registryPackage = federationAdminService.getRegistryObjectByRegistryId(registryId,
+                    sourceIds);
+        } catch (FederationAdminException e) {
+            String message = "Error getting registry package.";
+            LOGGER.error("{} For metacard id: '{}', optional sources: {}",
```
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Should this log say `metacard` id or `registry` id? Seems like you are actually logging the `registry` id.

File: catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpointTest.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    @Test
+    public void testPublish() throws Exception {
+        Response response = restEndpoint.publish(REGISTRY_ID, DESTINATION_ID);
+        verify(registryPublicationService).publish(REGISTRY_ID, DESTINATION_ID);
+        assertThat(response.getStatusInfo()
+                .getStatusCode(), is(Response.Status.OK.getStatusCode()));
+    }
+
+    @Test
+    public void testUnpublishMissingFirstArg() throws Exception {
```
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpointTest.java]: Recommend re-naming these two methods to something like `testUnpublishNullRegistryId`. Might be a little more clear.

File: catalog/spatial/registry/registry-rest-endpoint/pom.xml (hunk 5 above)
Comment (lessarderic) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: @shaundmorris: should we use `dependency-management` or a version property?

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        return Response.status(Response.Status.OK)
+                .build();
+    }
+
+    @Path("/{registryId}/publication/{sourceId}")
+    @DELETE
+    public Response unpublish(@PathParam("registryId") String registryId,
+            @PathParam("sourceId") String destinationId) {
+        if (StringUtils.isBlank(registryId) || StringUtils.isBlank(destinationId)) {
+            return Response.status(Response.Status.BAD_REQUEST)
```
Comment (lessarderic) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: We should be able to replace all those with the proper `javax.ws.rs` exceptions. See _Exception Summary_ here: http://docs.oracle.com/javaee/7/api/javax/ws/rs/package-summary.html.

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelper.java [hunk 11]
```diff
@@ -58,14 +41,7 @@
...
 
-/**
- * Generates a registry report for a given registry metacard.
- * Displays the Name, General, Collections,
- * Services, Organizations, and Contacts information for the node.
- */
-@Path("/")
-public class RegistryReportViewer {
-    private static final Logger LOGGER = LoggerFactory.getLogger(RegistryReportViewer.class);
+public class RegistryReportHelper {
```
Comment (lessarderic) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelper.java]: Can we name this class something else than `Helper`? Using _Helper_ in class name is considered a code smell and usually indicates that the class' responsibilities are unclear.

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            html = template.apply(registryMap);
+        } catch (IOException e) {
+            LOGGER.error("Error compiling and applying report template.");
+        }
+
+        return Response.ok(html)
+                .build();
+    }
+
+    public void setRegistryPublicationService(
```
Comment (lessarderic) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Any specific reason we're using setter injection over constructor injection?

File: catalog/spatial/registry/registry-rest-endpoint/pom.xml (hunk 5 above)
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: @lessarderic that is a great question. Although both dependency management and the version property work, it seems like dependency management is more appropriate and designed for multi-module cases. However, I like @coyotesqrl idea of bundle wrapping this one or going with the bundle 0 approach.

Comment (coyotesqrl) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: I think wrapping it is the best choice. It's easy enough to do and to upgrade later and matches what we've done with Guava.

File: catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpointTest.java (hunk 9 above)
Comment (figliold) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpointTest.java]: Same comment as @shaundmorris has.  There are also other instances of this in this file.

File: catalog/spatial/registry/registry-rest-endpoint/src/test/resources/csw-simplified-registry-package.xml [hunk 13]
```diff
@@ -33,6 +33,14 @@
     <rim:RegistryObjectList>
         <!-- defines the general node information -->
         <rim:ExtrinsicObject id="urn:registry:node" objectType="urn:registry:federation:node">
+            <rim:Description>
+                <rim:LocalizedString
+                        value="Extrinsic Object Description"/>
+            </rim:Description>
+            <rim:Name>
+                <rim:LocalizedString value="Extrinsic Object Name"/>
+            </rim:Name>
```
Comment (figliold) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/resources/csw-simplified-registry-package.xml]: Do we have to remember to update this version?

File: catalog/spatial/registry/registry-rest-endpoint/pom.xml (hunk 5 above)
Comment (shaundmorris) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: I ended up creating https://codice.atlassian.net/browse/DDF-2223 to capture the plan for commons-lang and I would expect that to be a separate PR across multiple modules.

File: catalog/spatial/registry/registry-rest-endpoint/pom.xml (hunk 2 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/pom.xml]: Added the schema bindings dependency

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 3 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Changed to INTERNAL_SERVER_ERROR

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 4 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Changed

File: catalog/spatial/registry/registry-rest-endpoint/src/test/resources/csw-simplified-registry-package.xml (hunk 13 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/resources/csw-simplified-registry-package.xml]: No, it's just a sample version in the sample xml. Has nothing to do with the version of DDF.

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 8 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Changed to registry id

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 12 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: No specific reason. Just personal preference. Haven't read anything concrete that details when to use argument injection vs property injection.

File: catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpointTest.java (hunk 9 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/test/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpointTest.java]: Renamed

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 6 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: They're both saying the same thing. Neither the registry id nor the destination id can be blank. The if checks if either is blank and the message says both cannot be blank. The check has been changed as suggested below.

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 7 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Extracted the logic.

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelper.java (hunk 11 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/report/RegistryReportHelper.java]: Renamed to RegistryReportMapBuilder

File: catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java (hunk 10 above)
Comment (gordocanchola) [inline catalog/spatial/registry/registry-rest-endpoint/src/main/java/org/codice/ddf/registry/rest/endpoint/RegistryRestEndpoint.java]: Tried using the proper javax.ws.rs exceptions but they didn't provide much value without implementing an ExceptionMapper. Reverted to returning the Response.

**Comment by shaundmorris** [comment]
@gordocanchola can you resolve the merge conflicts on this one?

**Comment by clockard** [comment]
Hero build succeeded but noticed an issue when manually testing. Looks like the service binding properties are not being displayed correctly. 
screen shot 2016-06-22 at 2 56 24 pm

**Comment by vinamartin** [comment]
@clockard If I had to guess, it looks like `Properties` and `Bindings` are being treated as bindings here.

**Comment by gordocanchola** [comment]
There must have been an issue merging the moved file in ddf-registry to what was in codice. Changed the file to match what was in master.

**Comment by clockard** [comment]
Re-heroed successfully. Looks good 👍
